pantry/src/pantry.rs

```diff
@@ -396,6 +400,116 @@ impl PantryEntry {
         self.volume.deactivate().await?;
         Ok(())
     }
+
+    pub async fn volume_is_active(&self) -> Result<bool, CrucibleError> {
+        self.volume.query_is_active().await
+    }
+
+    pub async fn activate(&self) -> Result<(), CrucibleError> {
+        self.volume.activate().await
+    }
+
+    pub async fn replace(
+        &self,
+        new_vcr: VolumeConstructionRequest,
+    ) -> Result<ReplaceResult, CrucibleError> {
+        let mut mg = self.volume_construction_request.lock().await;
+        let current_vcr = mg.clone();
+
+        let result = self
+            .volume
+            .target_replace(current_vcr, new_vcr.clone())
+            .await?;
+
+        // If target_replace returned Ok, then replace this entry's vcr with the
+        // new one.
+        *mg = new_vcr;
+
+        Ok(result)
+    }
+}
+
+#[derive(Default)]
+pub struct PantryJobs {
+    job_handles: BTreeMap<String, JoinHandle<Result<(), CrucibleError>>>,
+    volume_id_to_job_ids: BTreeMap<String, BTreeSet<String>>,
+}
+
+impl PantryJobs {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn total_job_handles(&self) -> usize {
+        self.job_handles.len()
+    }
+
+    pub fn num_job_handles_for_volume(&self, volume_id: &str) -> usize {
+        match self.volume_id_to_job_ids.get(volume_id) {
+            Some(job_ids) => job_ids.len(),
+            None => 0,
+        }
+    }
+
+    pub fn get(
+        &self,
+        job_id: &str,
+    ) -> Option<&JoinHandle<Result<(), CrucibleError>>> {
+        self.job_handles.get(job_id)
+    }
+
+    pub fn contains_job(&self, job_id: &str) -> bool {
+        self.job_handles.contains_key(job_id)
+    }
+
+    pub fn insert(
+        &mut self,
+        volume_id: String,
+        job_id: String,
+        handle: JoinHandle<Result<(), CrucibleError>>,
+    ) -> Option<JoinHandle<Result<(), CrucibleError>>> {
+        let inserted = self
+            .volume_id_to_job_ids
+            .entry(volume_id)
+            .or_default()
+            .insert(job_id.clone());
+
+        assert!(inserted);
+
+        self.job_handles.insert(job_id, handle)
+    }
+
+    pub fn remove(
+        &mut self,
+        job_id: &str,
+    ) -> Option<JoinHandle<Result<(), CrucibleError>>> {
+        // Does this job exist?
+        let job = self.job_handles.remove(job_id);
+
+        if job.is_some() {
+            // Each job id should map to one volume id
+            let mut job_volume_id: Vec<String> = self
+                .volume_id_to_job_ids
+                .iter()
+                .filter(|(_, job_ids)| job_ids.contains(job_id))
+                .map(|(volume_id, _)| volume_id.clone())
+                .collect();
+
+            assert_eq!(job_volume_id.len(), 1);
+
+            let volume_id = job_volume_id.pop().unwrap();
+
+            let existing = self
+                .volume_id_to_job_ids
```

Comment:
What is this doing here?  Removing a volume_id from the `volume_id_to_job_ids` list?

Reply:
right - remove the job from the volume's list of job ids.